FIXES: 7971 [2025-06 LWG Motion 34] P3552R3 Add a Coroutine Task (#8032)
cplusplus/draft: source/exec.tex

  // \ref{exec.with.awaitable.senders}
  template<@\exposconcept{class-type}@ Promise>
    struct with_awaitable_senders;

  // \ref{exec.affine.on}
  struct @\libglobal{affine_on_t}@ { @\unspec@ };
  inline constexpr affine_on_t @\libglobal{affine_on}@{};

  // \ref{exec.inline.scheduler}
  class @\libglobal{inline_scheduler}@;

  // \ref{exec.task.scheduler}
  class @\libglobal{task_scheduler}@;

  // \ref{exec.task}

>   template<class E>

Reviewer: As dietmar explained above, those definitions exist, but are missing declarations in the synopsis.

>   template<class E>

Reviewer 2: Ah, ok.  And those declarations seem to have no add-on info, so they probably should exist in the synopsis only.

>   template<class E>

Author: I have created a commit moving the definition of these two class templates to the synopsis (it is a [separate commit](https://github.com/cplusplus/draft/pull/8032/commits/1c45cffe76d86f395744ea12072e55a1e37e97db)).

return as_awaitable(std::forward<Value>(value), static_cast<Promise&>(*this));
\end{codeblock}
\end{itemdescr}

\rSec2[exec.affine.on]{\tcode{execution::affine_on}}

\pnum
\tcode{affine_on} adapts a sender into one that completes on
the specified scheduler. If the algorithm determines that the adapted
sender already completes on the correct scheduler it can avoid any
scheduling operation.

\pnum
The name \tcode{affine_on} denotes a pipeable sender adaptor
object. For subexpressions \tcode{sch} and \tcode{sndr}, if
\tcode{decltype((sch))} does not satisfy \libconcept{scheduler}, or
\tcode{decltype((sndr))} does not satisfy \libconcept{sender},
\tcode{affine_on(sndr, sch)} is ill-formed.

\pnum
Otherwise, the expression \tcode{affine_on(sndr, sch)} is
expression-equivalent to:
\begin{codeblock}
transform_sender(@\exposid{get-domain-early}@(sndr), @\exposid{make-sender}@(affine_on, sch, sndr))
\end{codeblock}
except that \tcode{sndr} is evalutated only once.

\pnum

> The exposition-only class template \exposid{impls-for}\iref{exec.snd.general}

Author: There were just two \irefs which didn't match references in the original text. I have removed them (with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/5df4a38d14d138778b3a80307410f96f8c34af13)).

  public:
    using scheduler_concept = scheduler_t;

    constexpr @\exposid{inline-sender}@ schedule() noexcept { return {}; }
    constexpr bool operator== (const inline_scheduler&) const noexcept = default;
  };
}
\end{codeblock}

\pnum
\tcode{inline_scheduler} is a class that models
\libconcept{scheduler}\iref{exec.sched}. All objects of type
\tcode{inline_scheduler} are equal.

\pnum
\exposid{inline-sender} is an exposition-only type that satisfies
\libconcept{sender}. The type
\tcode{completion_signatures_of_t<\exposid{inline-sender}>}
is \tcode{completion_signatures<set_value_t()>}.

\pnum
Let \tcode{sndr} be an expression of type \exposid{inline-sender},
let \tcode{rcvr} be an expression such that
\tcode{\libconcept{receiver_of}<decltype((rcvr)), CS>} is {true}
where \tcode{CS} is \tcode{completion_signatures<set_value_t()>},
then:

\begin{itemize}
\item the expression \tcode{connect(sndr, rcvr)} has
type \tcode{\exposid{inline-state}<remove_cvref_t<decltype((rcvr))>>}
and is potentially-throwing if and only if
\tcode{((void)sndr, auto(rcvr))} is potentially-throwing, and
\item the expression
\tcode{get_completion_scheduler<set_value_t>(get_env(sndr))} has

> type\linebreak\tcode{inline_scheduler} and is potentially-throwing % avoid Overfull

Author: I have remove all uses of `\linebreak` and addressed all overfull hboxes.

if and only if \tcode{get_env(sndr)} is potentially-throwing.

> \end{itemize}

Reviewer 2: I think other parts of senders-receivers do sndr and rcvr as code font; I'd be ok with trying to be consistent here.  But please check my assumption.

> \end{itemize}

Author: Based on looking at other sections of `exec.tex` I believe Jens is right: the function parameters use neither `\placeholder` nor `\exposid`. I'll leave it consistent unless told otherwise.

\pnum
Let \tcode{o} be a non-const lvalue of type
\tcode{\exposid{inline-state}<Rcvr>}, and let \tcode{REC(o)} be
a non-const lvalue reference to an object of type \tcode{Rcvr} that
was initialized with the expression \tcode{rcvr} passed to an
invocation of \tcode{connect} that returned \tcode{o}, then:

\begin{itemize}
\item the object to which \tcode{REC(o)} refers remains valid for
the lifetime of the object to which \tcode{o} refers, and
\item the expression \tcode{start(o)} is equivalent to
\tcode{set_value(std::move(REC(o)))}.
\end{itemize}

\rSec2[exec.task.scheduler]{\tcode{execution::task_scheduler}}

\begin{codeblock}
namespace std::execution {
  class task_scheduler {

>     class @\exposidnc{ts-sender}@; // \expos

Author: The CI over here complained that `sender` should reference the `sender`. However, `sender` is a local class name (which does model `sender`), i.e., it isn't a reference to the `sender` concept. If there is a different work-around which keeps the name `sender` for the exposition-only type I'm happy to use it.

>     class @\exposidnc{ts-sender}@; // \expos

Reviewer: Ok weird stuff, that's beyond my TeX expertise.

    template<@\libconcept{receiver}@ R>
    class @\exposidnc{state}@;  // \expos

  public:

>     using scheduler_concept = scheduler_t;

Author: You mean I should use something like `@\libglobal{scheduler_concept}@`? In other place in this section (mostly using `using sender_concept = sender_t;`) there isn't any additional annotation. `scheduler_concept` is a naming coming from the [`scheduler` concept](https://eel.is/c++draft/exec.sched).

      && @\libconcept{scheduler}@<Sch>
    friend bool operator== (const task_scheduler& lhs, const Sch& rhs) noexcept;

  private:
    shared_ptr<void> @\exposidnc{sch_}@; // \expos
  };
}
\end{codeblock}

\pnum
\tcode{task_scheduler} is a class that models
\libconcept{scheduler}\iref{exec.sched}.  Given on object \tcode{s}
of type \tcode{task_scheduler}, let \tcode{SCHED(s)} be the object
owned by \tcode{s.\exposid{sch_}}.

\begin{itemdecl}
template<class Sch, class Allocator = allocator<void>>
  requires(!@\libconcept{same_as}@<task_scheduler, remove_cvref_t<Sch>>) && @\libconcept{scheduler}@<Sch>
explicit task_scheduler(Sch&& sch, Allocator alloc = {});
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Initialize \exposid{sch_} with
\tcode{allocate_shared<remove_cvref_t<Sch>>(alloc, \linebreak % avoid Overfull
std::forward<Sch>(sch))}.

\pnum
\recommended Implementations should avoid the use of dynamically
allocated memory for small scheduler objects.

\pnum
\remarks
Any allocations performed by construction of \exposid{ts-sender} or
\exposid{state} objects resulting from calls on \tcode{*this} are
performed using a copy of \tcode{alloc}.
\end{itemdescr}

\begin{itemdecl}

> @\exposid{ts-sender}@ schedule();

Author: I have addressed the missing index members with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/9953567985c8c3720b2d0561ca8802505684f1a7).

    template<@\libconcept{receiver}@ Rcvr>
    class @\exposidnc{state}@; // \expos

  public:
    using sender_concept = sender_t;
    using completion_signatures = @\seebelow@;
    using allocator_type = @\seebelow@;
    using scheduler_type = @\seebelow@;
    using stop_source_type = @\seebelow@;
    using stop_token_type = decltype(declval<stop_source_type>().get_token());
    using error_types = @\seebelow@;

    // \ref{task.promise}
    class promise_type;

    task(task&&) noexcept;
    ~task();

    template<@\libconcept{receiver}@ Rcvr>
    @\exposid{state}@<Rcvr> connect(Rcvr&& rcvr);

  private:
    coroutine_handle<promise_type> @\exposidnc{handle}@; // \expos
  };
}
\end{codeblock}

\pnum
\tcode{task<T, E>} models \libconcept{sender}\iref{exec.snd}
if \tcode{T} is \tcode{void}, a reference type, or an \cv{}-unqualified
non-array object type and \tcode{E} is a class type. Otherwise a program
that instantiates the definition of \tcode{task<T, E>} is ill-formed.

\pnum
The nested types of \tcode{task} template specializations
are determined based on the \tcode{Environment} parameter:

\begin{itemize}
\item \tcode{allocator_type} is \tcode{Environment::allocator_type}

> if that qualified-id is valid and denotes a type, \tcode{allocator<byte>}

Author: I have wrapped `qualified_id` int `\grammarterm` with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/619f2ad92587408001f3d42db0dc1f3b66efcd1e).

if that qualified-id is valid and denotes a type,
\tcode{inplace_stop_source} otherwise.
\item \tcode{error_types} is \tcode{Environment::error_types} if
that qualified-id is valid and denotes a type, \linebreak % avoid Overfull
\tcode{completion_signatures<set_error_t(exception_ptr)>} otherwise.
\end{itemize}

\pnum
 A program is ill-formed if \tcode{error_types} is not a
specialization of \tcode{completion_signatures<ErrorSigs...>} or
\tcode{ErrorSigs} contains an element which is not of the form
\tcode{set_error_t(E)} for some type \tcode{E}.

\pnum
The type alias \tcode{completion_signatures} is a specialization
of \tcode{execution::completion_signatures} with the template
arguments (in unspecified order):

\begin{itemize}
\item \tcode{set_value_t()} if \tcode{T} is \tcode{void}, and
\tcode{set_value_t(T)} otherwise;
\item template arguments of the specialization of
\tcode{execution::completion_signatures} denoted by \tcode{error_types};
and
\item \tcode{set_stopped_t()}.
\end{itemize}

\pnum
\tcode{allocator_type} shall meet the \oldconcept{Allocator}
requirements\iref{allocator.requirements.general}.

\rSec3[task.members]{\tcode{task} Members}

> 

Reviewer: ```suggestion

\indexlibraryctor{task}%
```

Various other index entries missing in this area.

\begin{itemdecl}
task(task&& other) noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Initializes \exposid{handle} with \tcode{exchange(other.\exposid{handle},
\{\})}.
\end{itemdescr}

\begin{itemdecl}
~task();
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Equivalent to:

\begin{codeblock}
if (@\exposid{handle}@)
  @\exposid{handle}@.destroy();
\end{codeblock}
\end{itemdescr}

\begin{itemdecl}
template<@\libconcept{receiver}@ R>

> @\exposid{state}@<R> connect(R&& recv);

Reviewer: ```suggestion
  @\exposid{state}@<R> connect(R&& recv);
```

\end{itemdecl}

\begin{itemdescr}
\pnum
\expects
\tcode{bool(\exposid{handle})} is \tcode{true}.

\pnum
\effects
Equivalent to: \tcode{return \exposid{state}<R>(exchange(\exposid{handle},
\{\}), std::forward<R>(recv));}
\end{itemdescr}

\rSec3[task.state]{\tcode{Class template \tcode{task::state}}}

\begin{codeblock}
namespace std::execution {
  template<class T, class Environment>
    template<@\libconcept{receiver}@ Rcvr>
  class task<T, Environment>::@\exposid{state}@ { // \expos
  public:
    using operation_state_concept = operation_state_t;

    template<class R>

>     @\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);

Reviewer: ```suggestion
      @\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);
```

    ~@\exposid{state}@();
    void start() & noexcept;

private:
    using @\exposidnc{own-env-t}@ = @\seebelow@;     // \expos
    coroutine_handle<promise_type> @\exposidnc{handle}@;  // \expos
    remove_cvref_t<Rcvr>              @\exposidnc{rcvr}@;    // \expos
    @\exposid{own-env-t}@                      @\exposidnc{own-env}@; // \expos
    Environment                    @\exposidnc{environment}@; // \expos
  };
}
\end{codeblock}

\begin{itemdescr}
\pnum
The type \exposid{own-env-t} is \tcode{Environment::template

> env_type<decltype(get_env(\linebreak declval<Rcvr>()))>} if that % avoid Overfull

Reviewer: ```suggestion
env_type<decltype(get_env(\brk{}declval<Rcvr>()))>} if that
```

> qualified-id is valid and denotes a type, \tcode{env<>} otherwise.

Reviewer: ```suggestion
\grammarterm{qualified-id} is valid and denotes a type, \tcode{env<>} otherwise.
```

\end{itemdescr}

\begin{itemdecl}
template<class R>

> @\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);

Reviewer: ```suggestion
  @\exposid{state}@(coroutine_handle<promise_type> h, R&& rr);
```

\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Initializes

\item \exposid{handle} with \tcode{std::move(h)};
\item \exposid{rcvr} with \tcode{std::forward<R>(rr)};
\item \exposid{own-env}

>     with \tcode{\exposid{own-env-t}(get_env(\exposid{rcvr}))} if that expression

Reviewer: This indentation within bullets is wrong. Just get rid of it an make it all flat.

Author: I have removed the leading indentation with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/2ec36bb49f287d73cd17e8dcf9ebbb0f13d3ceed).

    is valid and \tcode{\exposid{own-env-t}()} otherwise.
    If neither of these expressions is valid, the program is ill-formed.
\item \exposid{environment} with \tcode{Environment(\exposid{own-env})} if that expression is
    valid, otherwise \tcode{Environment(get_env(\exposid{rcvr}))}
    if this expression is valid, otherwise \tcode{Environment()}.

>     If neither of these expressions is valid, the program is ill-formed.

Reviewer: There is no `itemize` environment surrounding these.

I'm surprised it even builds, but I guess that's a typical TeX moment.

\end{itemdescr}

\begin{itemdecl}
~@\exposid{state}@();
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Equivalent to:

> 

Reviewer: Wrong empty line.

Reviewer: Same issue in a number of other places, I won't make any more separate comments for this.

\begin{codeblock}
if (@\exposid{handle}@)
  @\exposid{handle}@.destroy();
\end{codeblock}
\end{itemdescr}

\begin{itemdecl}
void start() & noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
Let \tcode{prom} be the object \tcode{\exposid{handle}.promise()}.

> Associates \tcode{STATE(prom)}, \tcode{RCVR(prom)}, and \tcode{SCHED(prom)}

Reviewer: ```suggestion
Associates
\tcode{\exposid{STATE}(prom)},
\tcode{\exposid{RCVR}(prom)}, and
\tcode{\exposid{SCHED}(prom)}
```

with \tcode{*this} as follows:

\begin{itemize}
\item \tcode{STATE(prom)} is \tcode{*this}.
\item \tcode{RCVR(prom)} is \exposid{rcvr}.
\item \tcode{SCHED(prom)} is the object initialized
with \tcode{scheduler_type(get_scheduler(get_env(\exposid{rcvr})))}
if that expression is valid and \tcode{scheduler_type()} otherwise.
If neither of these expressions is valid, the program is ill-formed.
\end{itemize}

Let \tcode{st} be \tcode{get_stop_token(get_env(\exposid{rcvr}))}.
Initializes \tcode{prom.\exposid{token}} and
\tcode{prom.\exposid{source}} such that

\begin{itemize}
\item
\tcode{prom.\exposid{token}.stop_requested()} returns
\tcode{st.stop_requested()};
\item
\tcode{prom.\exposid{token}.stop_possible()} returns
\tcode{st.stop_possible()}; and
\item
for types \tcode{Fn} and \tcode{Init} such that both
\tcode{\libconcept{invocable}<Fn>} and
\tcode{\libconcept{constructible_from}<Fn, Init>} are modeled,
\tcode{stop_token_type::callback_type<Fn>} models

> \tcode{\exposid{stoppable-callback-for}<Fn,\linebreak stop_token_type, % avoid Overfull

Reviewer: ```suggestion
\tcode{\exposconcept{stoppable-callback-for}<Fn,\brk{} stop_token_type, Init>}.
```

Init>}.
\end{itemize}

After that invokes \tcode{\exposid{handle}.resume()}.
\end{itemdescr}

\rSec3[task.promise]{Class \tcode{task::promise_type}}

\begin{codeblock}
namespace std::execution {
  template<class E>
  struct with_error {
    using type = remove_cvref_t<E>;
    type error;
  };

>   template<class E>
>   with_error(E) -> with_error<E>;

Reviewer: ```suggestion
  template<class E>
    with_error(E) -> with_error<E>;
```
In general, there should be two spaces of indentation after a *template-head*, except for class definitions. I won't make any more comments for this.

Author: I applied the change with [this commit](https://github.com/cplusplus/draft/pull/8032/commits/70456ebe2b2681393bf3da89887c1c1e48356150).

  template<@\libconcept{scheduler}@ Sch>
  struct change_coroutine_scheduler {
    using type = remove_cvref_t<Sch>;
    type scheduler;
  };
  template<@\libconcept{scheduler}@ Sch>
  change_coroutine_scheduler(Sch) -> change_coroutine_scheduler<Sch>;

  template<class T, class Environment>
  class task<T, Environment>::promise_type {
  public:
    template<class... Args>
    promise_type(const Args&... args);

    task get_return_object() noexcept;

    auto initial_suspend() noexcept;
    auto final_suspend() noexcept;

    void uncaught_exception();
    coroutine_handle<> unhandled_stopped();

>     void return_void();             // present only if is_void_v<T> is true;
>     template<class V>
>     void return_value(V&& value);   // present only if is_void_v<T> is false;

Reviewer: ```suggestion
    void return_void();               // present only if \tcode{is_void_v<T>} is \tcode{true};
    template<class V>
      void return_value(V&& value);   // present only if \tcode{is_void_v<T>} is \tcode{false};
```

    template<class E>
    @\unspec@ yield_value(with_error<E> error);

    template<class A>
    auto await_transform(A&& a);
    template<class Sch>
    auto await_transform(change_coroutine_scheduler<Sch> sch);

    @\unspec@ get_env() const noexcept;

    template<class... Args>
    void* operator new(size_t size, Args&&... args);

    void operator delete(void* pointer, size_t size) noexcept;

  private:
    using @\exposidnc{error-variant}@ = @\seebelow@; // \expos

    allocator_type    @\exposidnc{alloc}@;  // \expos
    stop_source_type  @\exposidnc{source}@; // \expos
    stop_token_type   @\exposidnc{token}@;  // \expos

>     optional<T>       @\exposidnc{result}@; // \expos; present only if is_void_v<T> is false;

Reviewer: ```suggestion
    optional<T>       @\exposidnc{result}@; // \expos; present only if \tcode{is_void_v<T>} is \tcode{false};
```

    @\exposid{error-variant}@      @\exposidnc{errors}@; // \expos
  };
}
\end{codeblock}

\pnum
Let \tcode{prom} be an object of \tcode{promise_type} and let \tcode{tsk} be
the \tcode{task} object created by \tcode{prom.get_return_object()}. The
description below refers to objects \tcode{STATE(prom)},

> \tcode{RCVR(prom)}, and \tcode{SCHED(prom)} associated

Reviewer: Some more missing `\exposid` here.

with \tcode{tsk} during evalutation of \tcode{task::\exposid{state}<Rcvr>::start}
for some receiver \tcode{Rcvr}.

\pnum
\exposid{error-variant} is a \tcode{variant<monostate,
remove_cvref_t<E>...>}, with duplicate types removed, where \tcode{E...}
are template arguments of the specialization of
\tcode{execution::completion_signatures} denoted by
\tcode{error_types}.

\begin{itemdecl}
template<class... Args>
promise_type(const Args&... args);
\end{itemdecl}

\begin{itemdescr}
\pnum
\mandates
The first parameter of type \tcode{allocator_arg_t} (if any) is not
the last parameter.

\pnum
\effects
If \tcode{Args} contains an element of type \tcode{allocator_arg_t}
then \exposid{alloc} is initialized with the corresponding next
element of \tcode{args}. Otherwise, \exposid{alloc} is initialized
with \tcode{allocator_type()}.
\end{itemdescr}

\begin{itemdecl}
task get_return_object() noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\returns
A \tcode{task} object whose member \exposid{handle} is

> \tcode{coroutine_handle<promise_type>::\linebreak from_promise(*this)}. % avoid Overfull

Reviewer: ```suggestion
\tcode{coroutine_handle<promise_type>::\brk{}from_promise(*this)}.
```

\end{itemdescr}

\begin{itemdecl}
auto initial_suspend() noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\returns

> An awaitable object of unspecified type \iref{expr.await} whose

Reviewer: ```suggestion
An awaitable object of unspecified type\iref{expr.await} whose
```

member functions arrange for

\begin{itemize}
\item the calling coroutine to be suspended,
\item the coroutine to be resumed on an execution agent of the
execution resource associated with \tcode{SCHED(*this)}.
\end{itemize}
\end{itemdescr}

\begin{itemdecl}
auto final_suspend() noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\returns

> An awaitable object of unspecified type \iref{expr.await} whose

Reviewer: ```suggestion
An awaitable object of unspecified type\iref{expr.await} whose
```

member functions arrange for the completion of the asynchronous
operation associated with \tcode{STATE(*this)} by invoking:

\begin{itemize}
\item
    \tcode{set_error(std::move(RCVR(*this)), std::move(e))}
    if \tcode{\exposid{errors}.index()} is greater than zero and
    \tcode{e} is the value held by \exposid{errors}, otherwise
\item
    \tcode{set_value(std::move(RCVR(*this)))} if \tcode{is_void<T>} is \tcode{true},
    and otherwise
\item

>  \tcode{set_value(std::move(RCVR(*this)), *\exposid{result})}.

Reviewer: Remove indentation for bullets here.

\end{itemize}
\end{itemdescr}

\begin{itemdecl}
template<class Err>
auto yield_value(with_error<Err> err);
\end{itemdecl}

\begin{itemdescr}
\pnum
\mandates
\tcode{std::move(err.error)} is convertible to exactly one of the
\tcode{set_error_t} argument types of \tcode{error_types}. Let
\tcode{Cerr} be that type.

\pnum
\returns

> An awaitable object of unspecified type \iref{expr.await} whose

Reviewer: ```suggestion
An awaitable object of unspecified type\iref{expr.await} whose
```

member functions arrange for the calling coroutine to be suspended
and then completes the asynchronous operation associated with
\tcode{STATE(*this)} by invoking \tcode{set_error(std::move(RCVR(*this)),
Cerr(std::move(err.error)))}.
\end{itemdescr}

\begin{itemdecl}
template<@\libconcept{sender}@ Sender>
auto await_transform(Sender&& sndr) noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\returns

> If \tcode{same_as<inline_scheduler, scheduler_type>} is \tcode{true}

Reviewer: ```suggestion
If \tcode{@\libconcept{same_as}@<inline_scheduler, scheduler_type>} is \tcode{true}
```

> returns \tcode{as_awaitable(\linebreak std::forward <Sender>(sndr), % avoid Overfull
> *this)}; otherwise returns

Reviewer: ```suggestion
returns \tcode{as_awaitable(\brk{}std::forward<Sender>(sndr), *this)};
otherwise returns
```

> \tcode{as_awaitable(affine_on(\linebreak std::forward<Sender>(sndr), % avoid Overfull
> SCHED(*this)), *this)}.

Reviewer: ```suggestion
\tcode{as_awaitable(affine_on(\brk{}std::forward<Sender>(sndr), \exposid{SCHED}(*this)), *this)}.
```

\end{itemdescr}

\begin{itemdecl}
template<class Sch>
auto await_transform(change_coroutine_scheduler<Sch> sch) noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects

> Equivalent to: \tcode{returns await_transform(just(exchange(SCHED(*this),
> \linebreak scheduler_type(sch.scheduler))), *this);} % avoid Overfull

Reviewer: ```suggestion
Equivalent to:
\begin{codeblock}
return await_transform(just(exchange(SCHED(*this),  scheduler_type(sch.scheduler))), *this);
\end{codeblock}
```
Note the typo in `returns` (paper has this), but I think this doesn't really need a separate commit.

\end{itemdescr}

\begin{itemdecl}
void uncaught_exception();
\end{itemdecl}

\begin{itemdescr}
\pnum
\effects
If the signature \tcode{set_error_t(exception_ptr)} is not an element

> of \tcode{error_types}, calls \tcode{terminate()} \iref{except.terminate}.

Reviewer: ```suggestion
of \tcode{error_types}, calls \tcode{terminate()}\iref{except.terminate}.
```

\end{itemdescr}

\begin{itemdecl}
@\unspec@ get_env() const noexcept;
\end{itemdecl}

\begin{itemdescr}
\pnum
\returns
An object \tcode{env} such that queries are forwarded as follows:

\begin{itemize}
\item \tcode{env.query(get_scheduler)} returns \tcode{scheduler_type(SCHED(*this))}.
\item \tcode{env.query(get_allocator)} returns \exposid{alloc}.
\item \tcode{env.query(get_stop_token)} returns \exposid{token}.
\item For any other query \tcode{q} and arguments \tcode{a...} a
    call to \tcode{env.query(q, a...)} returns
    \tcode{STATE(*this).} \tcode{environment.query(q, a...)} if this expression
    is well-formed and \tcode{forwarding_query(q)} is well-formed and is \tcode{true}.  Otherwise
    \tcode{env.query(q, a...)} is ill-formed.
\end{itemize}
\end{itemdescr}

\begin{itemdecl}
template<class... Args>
void* operator new(size_t size, const Args&... args);
\end{itemdecl}

\begin{itemdescr}
\pnum
If there is no parameter with type \tcode{allocator_arg_t} then let
\tcode{alloc} be \tcode{Allocator()}.  Let \tcode{arg_next} be the
parameter following the first \tcode{allocator_arg_t} parameter (if
any) and let \tcode{alloc} be \tcode{Allocator(arg_next)}.  Then

> \tcode{PAlloc} is \tcode{allocator_traits<Allocator>::template
> rebind_alloc\linebreak<U>} where \tcode{U} is an unspecified type % avoid Overfull

Reviewer: ```suggestion
\tcode{PAlloc} is \tcode{allocator_traits<Allocator>::template
re\-bind_alloc\brk{}<U>} where \tcode{U} is an unspecified type
```

whose size and alignment are both \tcode{__STDCPP_DEFAULT_NEW_ALIGNMENT__}.

\pnum
\mandates

\begin{itemize}
\item The first parameter of type \tcode{allocator_arg_t} (if any) is not the last parameter.
\item \tcode{Allocator(arg_next)} is a valid expression if there is a parameter
   of type \tcode{allocator_arg_t}.
\item \tcode{allocator_traits<PAlloc>::pointer} is a pointer type.
\end{itemize}

\pnum
\effects
Initializes an allocator \tcode{palloc} of type \tcode{PAlloc} with
\tcode{alloc}. Uses \tcode{palloc} to allocate storage for the
smallest array of \tcode{U} sufficient to provide storage for a
coroutine state of size \tcode{size}, and unspecified additional
state necessary to ensure that \tcode{operator delete} can later
deallocate this memory block with an allocator equal to \tcode{palloc}.

> \pnum

Reviewer: ```suggestion

\pnum
```

Reviewer 2: Yes, pnum needs an empty line in front (or a \begin{itemdescr}, otherwise it's not functioning properly.

The storage referenced by \tcode{s} may be used by \tcode{*this}
as temporary storage for the duration of the operation launched by this call.
\end{itemdescr}

> \rSec2[exec.affine.on]{\tcode{execution::affine_on}}

Reviewer 3: @dietmarkuehl I'm assuming this material is meant to go under "Coroutine utilities" (which happened to be at the end of [exec] prior to this meeting), rather than under "Namespace system_context_replaceability".

Author: yes, that is correct.

Reviewer 3: No, I've got it. It was just that the edit instructions were not literally true ("at the end of [exec]" should instead of named the subclause at whose end you wanted it), so I mainly just want to confirm that I understood the intention.
